sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,388 @@
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == 0xFF);
+}
+
+static void wait_rx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (rxqd == 0);
+}
+
+static inline void tx_fifo_write8(const dif_spi_host_t *spi_host,
+                                  uintptr_t srcaddr) {
+  uint8_t *src = (uint8_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write8(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+static inline void tx_fifo_write32(const dif_spi_host_t *spi_host,
+                                   uintptr_t srcaddr) {
+  uint32_t *src = (uint32_t *)srcaddr;
+  wait_tx_fifo(spi_host);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET, *src);
+}
+
+void dif_spi_host_fifo_write(const dif_spi_host_t *spi_host, const void *src,
+                             uint16_t len) {
+  uintptr_t ptr = (uintptr_t)src;
+
+  // If the pointer starts mis-aligned, write until we are aligned.
+  while (ptr & 3 && len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+
+  // Write complete 32-bit words to the fifo.
+  while (len > 3) {
+    tx_fifo_write32(spi_host, ptr);
+    ptr += 4;
+    len -= 4;
+  }
+
+  // Clean up any leftover bytes.
+  while (len > 0) {
+    tx_fifo_write8(spi_host, ptr);
+    ptr += 1;
+    len -= 1;
+  }
+}
+
+typedef struct queue {
+  int32_t length;
+  union {
+    uint64_t qword;
+    uint32_t word[2];
+    uint8_t byte[8];
+  };
+} queue_t;
+
+static inline void enqueue_byte(queue_t *queue, uint8_t data) {
+  queue->byte[queue->length++] = data;
+}
+
+static inline void enqueue_word(queue_t *queue, uint32_t data) {
+  if (queue->length % sizeof(uint32_t) == 0) {
+    queue->word[queue->length / sizeof(uint32_t)] = data;
+    queue->length += 4;
+  } else {
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+    data >>= 8;
+    enqueue_byte(queue, (uint8_t)data);
+  }
+}
+
+static inline uint8_t dequeue_byte(queue_t *queue) {
+  uint8_t val = queue->byte[0];
+  queue->qword >>= 8;
+  queue->length -= 1;
+  return val;
+}
+
+static inline uint32_t dequeue_word(queue_t *queue) {
+  uint32_t val = queue->word[0];
+  queue->qword >>= 32;
+  queue->length -= 4;
+  return val;
+}
+
+void dif_spi_host_fifo_read(const dif_spi_host_t *spi_host, void *dst,
+                            uint16_t len) {
+  uintptr_t ptr = (uintptr_t)dst;
+  // We always have to read from the RXFIFO as a 32-bit word.  We use a
+  // two-word queue to handle destination and length mis-alignments.
+  queue_t queue = {
+      0,
```

Comment:
Drop the comma so clang-format turns it into `{0}`.

Reply:
Done.